File: scraper.py
import requests
from bs4 import BeautifulSoup
import json
import re

DEPARTMENTS = {
    "CSE": {"code": "cse", "id": "28"},
    "CSE (AI & ML)": {"code": "cseaiml", "id": "4232"},
    "CSE (DS)": {"code": "cseds", "id": "4234"},
    "CSE (CS)": {"code": "csecs", "id": "4233"},
    "IT": {"code": "it", "id": "29"},
    "ECE": {"code": "ece", "id": "31"},
    "EEE": {"code": "eee", "id": "32"},
    "ME": {"code": "me", "id": "33"},
    "CE": {"code": "ce", "id": "34"},
    "AE": {"code": "ae", "id": "35"}
}

all_faculty = []
all_subjects = []

for dept_name, info in DEPARTMENTS.items():
    dept_code = info["code"]
    dept_id = info["id"]
    
    # 1. Scrape Faculty
    fac_url = f"https://www.iare.ac.in/?q=departmentlist/{dept_id}"
    print(f"Scraping Faculty: {fac_url}")
    try:
        res = requests.get(fac_url, timeout=10)
        soup = BeautifulSoup(res.text, 'html.parser')
        
        # In the HTML, faculty names are usually in <h3> tags under the main content
        main_content = soup.find(id="main-content") or soup
        h3s = main_content.find_all('h3')
        
        fac_count = 1
        for h3 in h3s:
            name = h3.get_text(strip=True)
            if not name or name in ['Outcome Based Education (OBE)', 'Handbook - Computer Science and Engineering', 'Governance', 'Facilities', 'Student Research Initiative', 'NIRF Data', 'Helpful Links', 'Infrastructure']:
                continue
            
            # Simple heuristic to check if it's a faculty name
            if name.startswith('Dr.') or name.startswith('Mr') or name.startswith('Ms'):
                fac_id = f"FAC-{dept_code.upper()}-{fac_count:03d}"
                all_faculty.append({
                    "id": fac_id,
                    "username": fac_id,
                    "password": fac_id,
                    "role": "Faculty",
                    "displayName": name,
                    "department": dept_name,
                    "firstLogin": True
                })
                fac_count += 1
    except Exception as e:
        print(f"Failed to scrape faculty for {dept_name}: {e}")
        
    # 2. Scrape Subjects (BT25 only)
    sub_url = f"https://www.iare.ac.in/?q=pages/btech-course-syllabi-bt25-{dept_code}"
    print(f"Scraping Subjects: {sub_url}")
    try:
        res = requests.get(sub_url, timeout=10)
        soup = BeautifulSoup(res.text, 'html.parser')
        
        # Look for links ending with .pdf and starting with a course code
        links = soup.find_all('a', href=re.compile(r'\.pdf$'))
        
        for link in links:
            txt = link.get_text(strip=True)
            # Match "CODE : Name" or just take the text if it has a hyphen/colon
            m = re.match(r'([A-Z0-9]+)\s*[:\-]\s*(.*)', txt)
            if m:
                code = m.group(1).strip()
                name = m.group(2).strip()
                
                # Check if we already added it
                if not any(s['name'] == name for s in all_subjects):
                    all_subjects.append({
                        "code": code,
                        "name": name,
                        "department": dept_name,
                        "modules": [1, 2, 3, 4, 5]
                    })
    except Exception as e:
        print(f"Failed to scrape subjects for {dept_name}: {e}")

output = {
    "faculty": all_faculty,
    "subjects": all_subjects,
    "departments": list(DEPARTMENTS.keys())
}

with open("src/utils/scraped_data.json", "w") as f:
    json.dump(output, f, indent=2)

print("Scraping complete! Saved to src/utils/scraped_data.json")

File: scraper.mjs
import fs from 'fs';
import https from 'https';

const DEPARTMENTS = {
  "CSE": { code: "cse", id: "28" },
  "CSE (AI & ML)": { code: "cseaiml", id: "113" },
  "CSE (DS)": { code: "cseds", id: "116" },
  "CSE (CS)": { code: "csecs", id: "115" },
  "IT": { code: "it", id: "27" },
  "ECE": { code: "ece", id: "29" },
  "EEE": { code: "eee", id: "30" },
  "ME": { code: "me", id: "31" },
  "CE": { code: "ce", id: "32" },
  "AE": { code: "ae", id: "26" }
};

function fetchUrl(url) {
  return new Promise((resolve, reject) => {
    https.get(url, (res) => {
      let data = '';
      res.on('data', chunk => data += chunk);
      res.on('end', () => resolve(data));
    }).on('error', reject);
  });
}

// Small delay to avoid hammering the server
function delay(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

async function scrape() {
  const allFaculty = [];
  const allSubjects = [];
  const allDeps = Object.keys(DEPARTMENTS);

  console.log("Starting scraping...");

  for (const [deptName, info] of Object.entries(DEPARTMENTS)) {
    const deptCode = info.code;
    
    // Scrape Faculty
    try {
      console.log(`Scraping faculty for ${deptName}...`);
      const html = await fetchUrl(`https://www.iare.ac.in/?q=departmentlist/${info.id}`);
      
      // Each faculty entry is inside a facInfo div with this structure:
      //   <div class="facInfo ...">
      //     <div class="facImg">
      //       <img src="...IARE10556_0.png..." />
      //     </div>
      //     <div class="facRight">
      //       <h3>Mr. C Praveen kumar </h3>
      //     </div>
      //   </div>
      // We match the image IARE ID and the h3 name within the same facInfo block.
      const facInfoRegex = /<div class="facInfo[^"]*">([\s\S]*?)<\/div>\s*<\/div>\s*<\/div>/g;
      let match;
      let count = 1;
      
      while ((match = facInfoRegex.exec(html)) !== null) {
        const block = match[1];
        
        // Extract IARE ID from image src
        const imgMatch = block.match(/IARE(\d+)/i);
        // Extract name from h3
        const nameMatch = block.match(/<h3>([^<]+)<\/h3>/);
        
        if (nameMatch) {
          let name = nameMatch[1].trim().replace(/&amp;/g, '&');
          
          if (name && (name.startsWith('Dr.') || name.startsWith('Mr') || name.startsWith('Ms') || name.startsWith('Prof'))) {
            const facId = imgMatch ? `IARE ${imgMatch[1]}` : `FAC-${deptCode.toUpperCase()}-${String(count).padStart(3, '0')}`;
            
            allFaculty.push({
              id: facId,
              username: facId,
              password: facId,
              role: "Faculty",
              displayName: name,
              department: deptName,
              firstLogin: true
            });
            
            process.stdout.write(`    ${name} => ${facId}\n`);
            count++;
          }
        }
      }
      
      console.log(`  Completed ${deptName}: ${count - 1} faculty`);
    } catch (e) {
      console.error(`Error scraping faculty for ${deptName}:`, e);
    }

    // Scrape Subjects
    try {
      console.log(`Scraping subjects for ${deptName}...`);
      const html = await fetchUrl(`https://www.iare.ac.in/?q=pages/btech-course-syllabi-bt25-${info.code}`);
      // Match <a> tags linking to BT25 PDF files, allowing nested HTML inside
      const linkRegex = /<a[^>]*href="[^"]*BT25[^"]*\.pdf"[^>]*>([\s\S]*?)<\/a>/gi;
      let match;
      while ((match = linkRegex.exec(html)) !== null) {
        // Strip any inner HTML tags (e.g. <span>) and decode entities
        let txt = match[1].replace(/<[^>]+>/g, '').trim().replace(/&amp;/g, '&').replace(/\s+/g, ' ');
        // Match subject code (4-6 alphanumeric chars) followed by colon/dash and subject name
        const m = txt.match(/^([A-Z]{3,5}\d{2,3})\s*[:\-–]\s*(.+)$/i);
        if (m) {
          const code = m[1].trim().toUpperCase();
          const name = m[2].trim();
          
          if (!allSubjects.some(s => s.code === code)) {
            allSubjects.push({
              code,
              name,
              department: deptName,
              modules: [1, 2, 3, 4, 5]
            });
          }
        }
      }
      console.log(`  Found ${allSubjects.filter(s => s.department === deptName).length} subjects for ${deptName}`);
    } catch (e) {
      console.error(`Error scraping subjects for ${deptName}:`, e);
    }
  }

  // Create generatedData.ts
  const tsContent = `// Auto-generated data from IARE website
export interface GeneratedSubject {
  code: string;
  name: string;
  department: string;
  modules: number[];
}

export const GENERATED_FACULTY = ${JSON.stringify(allFaculty, null, 2)};
export const GENERATED_SUBJECTS: GeneratedSubject[] = ${JSON.stringify(allSubjects, null, 2)};
export const GENERATED_DEPARTMENTS = ${JSON.stringify(allDeps, null, 2)};
`;

  fs.writeFileSync('src/utils/generatedData.ts', tsContent);
  console.log(`\nData successfully saved to src/utils/generatedData.ts!`);
  console.log(`Total: ${allFaculty.length} faculty, ${allSubjects.length} subjects, ${allDeps.length} departments`);
}

scrape();
